pkg/xds/experimental/client/client.go

```diff
@@ -0,0 +1,621 @@
+	delta discoverypb.AggregatedDiscoveryService_DeltaAggregatedResourcesClient,
+) error {
+	if err := c.sendInitialDiscoveryRequests(ctx, stream, delta); err != nil {
+		return fmt.Errorf("start request routine: %w", err)
+	}
+
+	errRespCh := make(chan error, 1)
+	go c.fetchResponses(ctx, errRespCh, stream, delta)
+	errLoopCh := make(chan error, 1)
+	go c.loop(ctx, errLoopCh, stream, delta)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case err, ok := <-errRespCh:
+			if !ok {
+				return fmt.Errorf("process responses: terminated")
+			}
+			if c.isStreamRetriableErr(err) {
+				continue
+			}
+			return fmt.Errorf("process responses: %w", err)
+		case err, ok := <-errLoopCh:
+			if !ok {
+				return fmt.Errorf("process loop: terminated")
+			}
+			if c.isStreamRetriableErr(err) {
+				continue
+			}
+			return fmt.Errorf("process loop: %w", err)
+		}
+	}
+}
+
+func (c *Client) isStreamRetriableErr(err error) bool {
+	if errors.Is(err, io.EOF) {
+		return false
+	}
+	return c.opts.RetryGrpcError(status.Code(err))
+}
+
+// sendInitialDiscoveryRequests sends requests for all configured ObservedResources.
+// It returns error only when a goroutine wasn't started yet.
+func (c *Client) sendInitialDiscoveryRequests(ctx context.Context,
+	stream discoverypb.AggregatedDiscoveryService_StreamAggregatedResourcesClient,
+	delta discoverypb.AggregatedDiscoveryService_DeltaAggregatedResourcesClient,
+) error {
+	log := c.log.With(logFeatureKey, "start-requests")
+	for _, typeUrl := range c.opts.BootstrapResources {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		var err error
+		if c.opts.UseSOTW {
+			req := &discoverypb.DiscoveryRequest{
+				Node:          c.node,
+				TypeUrl:       typeUrl,
+				ResourceNames: []string{},
+			}
+			log.Debug("Stream send", "req", req)
+			err = stream.Send(req)
+		} else {
+			req := &discoverypb.DeltaDiscoveryRequest{
+				Node:                   c.node,
+				TypeUrl:                typeUrl,
+				ResourceNamesSubscribe: []string{},
+			}
+			log.Debug("Delta send", "req", req)
+			err = delta.Send(req)
+		}
+		if err != nil {
+			return fmt.Errorf("initial requests: stream send: %w", err)
+		}
+	}
+	return nil
+}
+
+// fetchResponses will pass messages from Recv() calls to queue until Context ctx is done.
+func (c *Client) fetchResponses(ctx context.Context, errCh chan error,
+	stream discoverypb.AggregatedDiscoveryService_StreamAggregatedResourcesClient,
+	delta discoverypb.AggregatedDiscoveryService_DeltaAggregatedResourcesClient,
+) {
+	defer close(errCh)
+	log := c.log.With(logFeatureKey, "fetch-responses")
+	backoff := backoff.Exponential{
+		Min:        c.opts.MinBackoff,
+		Max:        c.opts.MaxBackoff,
+		ResetAfter: c.opts.BackoffReset,
+		Jitter:     true,
+		Name:       "xds-client-fetch-responses",
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+		if c.opts.UseSOTW {
```

Comment:
I had to visually diff between the if-blocks and it seems like the code is the same (or sublty the same, which is a problem) between the two apart from which object Recv() is called on and which response queue is used.

If the code is really the same, I suggest abstracting the delta vs state of the world using an interface and wrapper and then keep most of the if opts.UseSOTW out of the main code paths.